borsh-derive/src/lib.rs

```diff
@@ -1,25 +1,45 @@
 // TODO: re-enable this lint when we bump msrv to 1.58
 #![allow(clippy::uninlined_format_args)]
 extern crate proc_macro;
+use borsh_derive_internal::*;
+use borsh_schema_derive_internal::*;
+use mattro::MacroAttribute;
 use proc_macro::TokenStream;
 use proc_macro2::Span;
 use proc_macro_crate::crate_name;
-use syn::{Ident, ItemEnum, ItemStruct, ItemUnion};
-
-use borsh_derive_internal::*;
-use borsh_schema_derive_internal::*;
+use quote::quote;
+use syn::Attribute;
+use syn::{parse_macro_input, parse_quote, DeriveInput, Ident, ItemEnum, ItemStruct, ItemUnion};
 
-#[proc_macro_derive(BorshSerialize, attributes(borsh_skip))]
+#[proc_macro_derive(BorshSerialize, attributes(borsh_skip, use_discriminant))]
 pub fn borsh_serialize(input: TokenStream) -> TokenStream {
     let cratename = Ident::new(
         &crate_name("borsh").unwrap_or_else(|_| "borsh".to_string()),
         Span::call_site(),
     );
 
+    let for_derive_input = input.clone();
+    let derive_input = parse_macro_input!(for_derive_input as DeriveInput);
+
+    // Read the additional data
+    let mut use_discriminant = None;
+    for attr in &derive_input.attrs {
+        if attr.path.is_ident("use_discriminant") {
+            for token in attr.tokens.clone() {
+                if token.to_string() == "true" {
+                    use_discriminant = Some(true);
+                }
+                if token.to_string() == "false" {
+                    use_discriminant = Some(false);
+                }
```

Comment:
NOTE: Consider changing the function return value to `Result` as otherwise the `return` bellow is going to cause compilation error:

```suggestion
                use_discriminant = match token.to_string().as_str() {
                    "true" => Some(true),
                    "false" => Some(false),
                    _ => {
                        return Err(syn::Error::new(
                            Span::call_site(),
                            "`use_discriminant` only accepts true or false",
                        ));
                    }
                };
```